arch/risc-v/src/bl602/bl602_os_hal.c

```diff
@@ -0,0 +1,1035 @@
+                         len, prio);
+      if (ret < 0)
+        {
+          wlerr("ERROR: Failed to send message to mqueue error=%d\n",
+                ret);
+        }
+    }
+  else
+    {
+      ret = clock_gettime(CLOCK_REALTIME, &timeout);
+      if (ret < 0)
+        {
+          wlerr("ERROR: Failed to get time\n");
+          return false;
+        }
+
+      if (ticks)
+        {
+          bl_os_update_time(&timeout, ticks);
+        }
+
+      ret = file_mq_timedsend(&mq_adpt->mq, (const char *)item,
+                              len, prio, &timeout);
+      if (ret < 0)
+        {
+          wlerr("ERROR: Failed to timedsend message to mqueue error=%d\n",
+                ret);
+        }
+    }
+
+  return bl_os_errno_trans(ret);
+}
+
+/****************************************************************************
+ * Name: bl_os_mq_send
+ *
+ * Description:
+ *   Send message of low priority to queue within a certain period of time
+ *
+ * Input Parameters:
+ *   queue - Message queue data pointer
+ *   item  - Message data pointer
+ *   ticks - Wait ticks
+ *
+ * Returned Value:
+ *   True if success or false if fail
+ *
+ ****************************************************************************/
+
+int32_t bl_os_mq_send(void *queue, void *item, size_t len, uint32_t ticks)
+{
+  return bl_os_mq_send_generic(queue, item, len, ticks, 0);
+}
+
+/****************************************************************************
+ * Name: bl_os_mq_recv
+ *
+ * Description:
+ *   Receive message from queue within a certain period of time
+ *
+ * Input Parameters:
+ *   queue - Message queue data pointer
+ *   item  - Message data pointer
+ *   ticks - Wait ticks
+ *
+ * Returned Value:
+ *   True if success or false if fail
+ *
+ ****************************************************************************/
+
+int bl_os_mq_recv(void *queue, void *item, uint32_t ticks)
+{
+  ssize_t ret;
+  struct timespec timeout;
+  unsigned int prio;
+  struct mq_adpt *mq_adpt = (struct mq_adpt *)queue;
+
+  ret = clock_gettime(CLOCK_REALTIME, &timeout);
+
+  if (ret < 0)
+    {
+      wlerr("ERROR: Failed to get time\n");
+      return false;
+    }
+
+  if (ticks)
+    {
+      bl_os_update_time(&timeout, ticks);
+    }
+
+  ret = file_mq_timedreceive(&mq_adpt->mq,
+                             (char *)item,
+                             mq_adpt->msgsize,
+                             &prio,
+                             &timeout);
+  if (ret < 0)
+    {
+      wlerr("ERROR: Failed to timedreceive from mqueue error=%d\n", ret);
+    }
+
+  return ret > 0 ? true : false;
+}
+
+/****************************************************************************
+ * Name: bl_os_timer_create
+ *
+ * Description:
+ *
+ * Input Parameters:
+ *
+ * Returned Value:
+ *
+ ****************************************************************************/
+
+void *bl_os_timer_create(void *func, void *argv)
+{
+  void *timerid;
```

Comment:
kernel should use wdog or work with delay

Reply:
Is timer api only used in userspace?or for other reasons

Reply:
It isn't safe to use timer api in the kernel space.

Reply:
If you call the API designed for userspace from kernel component, you may hit some mysterious bugs.

Reply:
> If you call the API designed for userspace from kernel component, you may hit some mysterious bugs.

Got it, thanks